actionpack/test/controller/request_forgery_protection_test.rb

```diff
@@ -304,6 +304,50 @@ def test_should_allow_put_with_token_in_header
     assert_not_blocked { put :index }
   end
 
+  def test_should_allow_post_with_origin_checking_and_correct_origin
+    ActionController::Base.forgery_protection_origin_check = true
+    begin
+      session[:_csrf_token] = @token
+      @controller.stub :form_authenticity_token, @token do
+        assert_not_blocked do
+          @request.set_header 'HTTP_ORIGIN', 'http://test.host'
+          post :index, params: { custom_authenticity_token: @token }
+        end
+      end
+    ensure
+      ActionController::Base.forgery_protection_origin_check = false
+    end
+  end
+
+  def test_should_allow_post_with_origin_checking_and_no_origin
+    ActionController::Base.forgery_protection_origin_check = true
+    begin
+      session[:_csrf_token] = @token
+      @controller.stub :form_authenticity_token, @token do
+        assert_not_blocked do
+          post :index, params: { custom_authenticity_token: @token }
+        end
+      end
+    ensure
+      ActionController::Base.forgery_protection_origin_check = false
+    end
+  end
+
+  def test_should_block_post_with_origin_checking_and_wrong_origin
+    ActionController::Base.forgery_protection_origin_check = true
+    begin
+      session[:_csrf_token] = @token
+      @controller.stub :form_authenticity_token, @token do
+        assert_blocked do
+          @request.set_header 'HTTP_ORIGIN', 'http://bad.host'
+          post :index, params: { custom_authenticity_token: @token }
+        end
+      end
+    ensure
+      ActionController::Base.forgery_protection_origin_check = false
+    end
+  end
+
```

Comment:
> Prob worth testing some corner cases of Origin matching: case insensitivity, differing schemes and ports, ?

Given that we're just doing a string comparison, I'm not sure how meaningful corner-case tests would be. Testing those cases with actual browsers would be more useful, but I'm not sure that we can automate it.